output_derive/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+//! Custom derives for `output.rs`
+
+#![recursion_limit = "1024"]
+
+extern crate proc_macro;
+#[macro_use]
+extern crate quote;
+#[macro_use]
+extern crate syn;
+
+use proc_macro::TokenStream;
+use syn::{Data, DeriveInput, Fields};
+
+#[proc_macro_derive(RenderOutput)]
+pub fn render_output(input: TokenStream) -> TokenStream {
+    let ast: DeriveInput = parse_macro_input!(input as DeriveInput);
+
+    let name = &ast.ident;
+    let render_span = match ast.data {
+        Data::Struct(s) => match s.fields {
+            Fields::Named(..) => {
+                let fields = s.fields.iter().map(|f| {
+                    let x = f.ident.clone().unwrap();
+                    quote!(#x)
+                });
+                let names = s
+                    .fields
+                    .iter()
+                    .map(|f| f.ident.clone().unwrap().to_string());
+                quote! {
+                    let mut span = output::components::span();
+                    #(
+                        span = span.add_item(#names);
+                        span = span.add_item(": ");
+                        span = span.add_item(output::components::text(&self.#fields.to_string()));
+                        span = span.add_item("\n");
+                    )*
+                    span.render_for_humans(fmt)?;
+                }
+            }
+            Fields::Unnamed(..) => {
+                let field_count = s.fields.iter().count();
+                let fields = (0..field_count)
+                        .fold(Vec::new(), |mut res, i| {
+                            res.push(quote! { span = span.add_item(output::components::text(&self.#i.to_string())); });
+                            if i < field_count - 1 {
+                                res.push(quote! { span = span.add_item(", "); });
```

Comment:
Didn't I ask you recently about some clever function in `Vec` or something to do this? :sweat_smile:

Reply:
Yes and now I am confused: If I call `.to_string` above, I can totally put the fields in a vector and call `.join`. But I think I actually want to not render everything using `text` -- this would also require all fields to impl `Display`. I'll fiddle with this some more!

Reply:
That would definitely be prettier :grimacing:

Reply:
Ah, I remember: That would mean I'd have to actually collect the fields in a vector/slice, since there is no `join` on iterators in `std`.

Re Display: I don't want to add a bunch of impls of Render for various primitives, so it's gonna stay for now. I might open an issue about this later.